examples/gno.land/p/demo/grc/grc721/grc721_royalty.gno

```diff
@@ -0,0 +1,78 @@
+package grc721
+
+import (
+	"std"
+
+	"gno.land/p/demo/avl"
+)
+
+// royaltyNFT represents a non-fungible token (NFT) with royalty functionality.
+type royaltyNFT struct {
+	*metadataNFT          // Embedding metadataNFT for NFT functionality
+	tokenRoyaltyInfo avl.Tree  // AVL tree to store royalty information for each token
+}
+
+// Ensure that royaltyNFT implements the IGRC2981 interface.
+var _ IGRC2981 = (*royaltyNFT)(nil)
+
+// NewNFTWithRoyalty creates a new royalty NFT with the specified name, symbol, and royalty calculator.
+func NewNFTWithRoyalty(name string, symbol string) *royaltyNFT {
+	// Create a new NFT with metadata
+	nft := NewNFTWithMetadata(name, symbol)
+
+	return &royaltyNFT{
+		metadataNFT:        nft,
+		tokenRoyaltyInfo:   avl.Tree{},
+	}
+}
+
+// SetTokenRoyalty sets the royalty information for a specific token ID.
+func (r *royaltyNFT) SetTokenRoyalty(tid TokenID, royaltyInfo RoyaltyInfo) (bool, error) {
+	// Validate the payment address
+	if err := isValidAddress(royaltyInfo.PaymentAddress); err != nil {
+		return false, ErrInvalidRoyaltyPaymentAddress
+	}
+
+	// Check if royalty percentage exceeds 100%
+	if royaltyInfo.Percentage > 100 {
+		return false, ErrInvalidRoyaltyPercentage
+	}
+
+	// Check if the caller is the owner of the token
+	owner, err := r.metadataNFT.OwnerOf(tid)
+	if err != nil {
+		return false, err
+	}
+	caller := std.PrevRealm().Addr()
+	if caller != owner {
+		return false, ErrCallerIsNotOwner
+	}
+
+	// Set royalty information for the token
+	r.tokenRoyaltyInfo.Set(string(tid), royaltyInfo)
+	return true, nil
+}
+
+// RoyaltyInfo returns the royalty information for the given token ID and sale price.
+func (r *royaltyNFT) RoyaltyInfo(tid TokenID, salePrice uint64) (std.Address, uint64, error) {
```

Comment:
As in the interface, just return `RoyaltyInfo`

Reply:
same as above